src/util/parallel/data.ml

```diff
@@ -0,0 +1,242 @@
+open Batteries
+
+module type DefaultType = sig
+  type t
+  val default: unit -> t
+  val to_string: t -> string
+end
+
+module ConcurrentBucket (Key: Hashtbl.HashedType) (Val: DefaultType) = struct
+  type t = {
+    key: Key.t;
+    value: Val.t Atomic.t;
+    next: t option Atomic.t;
+  }
+
+  let create key = {
+    key = key;
+    value = Atomic.make @@ Val.default ();
+    next = Atomic.make None;
+  }
+
+  let create_with_value key value = {
+    key = key;
+    value = value;
+    next = Atomic.make None;
+  }
+
+  let create_with_value_and_next key value next = {
+    key = key;
+    value = value;
+    next = Atomic.make (Some next);
+  }
+
+  let find_option sll key =
+    let rec find sll key =
+      if Key.equal sll.key key then Some sll.value
+      else match Atomic.get sll.next with
+        | None -> None
+        | Some next -> find next key
+    in
+    find sll key
+
+  let find sll key =
+    match find_option sll key with
+    | None -> failwith "Key not found"
+    | Some value -> value
+
+  let rec find_create sll key =
+    if Key.equal sll.key key then (sll.value, false)
+    else (
+      match Atomic.get sll.next with
+      | None ->
+        let new_sll = {
+          key = key;
+          value = Atomic.make @@ Val.default (); 
+          next = Atomic.make None;
+        } in
+        let success = Atomic.compare_and_set sll.next None (Some new_sll) in
+        if success then (new_sll.value, true)
+        else find_create sll key 
+      | Some next -> find_create next key 
+    )
+
+  let rec insert_value sll key value =
+    if Key.equal sll.key key then ()
+    else (
+      match Atomic.get sll.next with
+      | None ->
+        let new_sll = {
+          key = key;
+          value = value;
+          next = Atomic.make None;
+        } in
+        let success = Atomic.compare_and_set sll.next None (Some new_sll) in
+        if not success then insert_value sll key value
+      | Some next -> insert_value next key value
+    )
+
+  let to_list sll =
+    let rec aux acc sll =
+      match sll with
+      | None -> acc
+      | Some sll ->
+        aux (acc @ [(sll.key, sll.value)]) (Atomic.get sll.next)
```

Comment:
Is the order relevant? Otherwise we could do `::`?

Reply:
Since a bucket is unsorted by nature, we can do `::`.